nixos/user-groups: add a toggle for user account creation

NotAShelf opened the pull request:
Microscopic change that allows users to toggle user accounts, per user, conditionally
under `users.users.<name>.enable`. If set to false, the user account will not be
created.

Allows for user accounts to be disabled in-line (similar to systemd services)
with minimum effort and itches a very specific scratch that I had, which I imagine
others might have as well. 

A similar mechanism can be implemented for groups as well, but I have omitted
such a change until I receive feedback.


<!--
^ Please summarise the changes you have done and explain why they are necessary here ^

For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [25.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) (or backporting [24.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) and [25.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

NotAShelf committed:
nixos/user-groups: add a toggle for user account creation

Microscopic change that allows users to toggle user accounts, per user, conditionally.

GetPsyched reviewed (approved):
Diff LGTM

aanderse commented:
awesome! i am pretty sure i will make use of this at some point, thanks

let's add a nixos test to ensure this functionality remains working in the future - sound good?

NotAShelf commented:
I'm happy to add a test, was waiting to get a review before I actually committed to writing it.

What should we test for? `enable -> user is created` and `disable -> user is created`?

aanderse commented:
> was waiting to get a review before I actually committed to writing it.

that makes sense to me :+1:


> What should we test for? `enable -> user is created` and `disable -> user is created anyway`?

yeah, i don't think much needs to be checked here, just that the basics are working

NotAShelf commented:
Sounds good, I'll add a test shortly.

NotAShelf committed:
nixos/tests: add user-enable-option

Specifically for checking if a user exist when `users.users.<name>.enable` is set to `true`.

NotAShelf commented:
Added a basic check to verify user with `id`. Grepping `/etc/passwd` would also work, but this is shorter :)

Checks if the user exists, and if it can log in. Should be sufficient for most cases, but I think we can also verify if `enable = false` works as intended.

NotAShelf commented:
Good thing we decided to test for that, I caught a regression introduced by the `mkIf` that would only appear if the user was disabled (which I had not yet tested for in my configuration)

NotAShelf commented:
Pushed a fix, we now filter `cfg.users` for enabled users in `spec`, which is passed to the perl script.

NotAShelf commented:
Also added test cases for system users, I'll apply a workaround and write testcases for userborn shortly.

On that note, does ofborg eval all on module changes, or do we need to hope edge cases get caught by users?

NotAShelf committed:
nixos/userborn: filter enabled users

NotAShelf commented:
And that settles userborn, I think. Maybe @nikstur would like to propose more tests for userborn.

NotAShelf commented:
Looks like the formatting CI is failing, should I format `nixos/modules/services/system/userborn.nix?

eclairevoyant reviewed (commented):
> should I format `nixos/modules/services/system/userborn.nix?

Yes, since it was formatted prior to this PR.

NotAShelf commented:
I detest this new formatting, but there we go.

aanderse commented:
@nikstur do you mind taking a look at this? i would love to get this merged

---

also, anyone interested in this PR may also be interested in another [similar PR](https://github.com/NixOS/nixpkgs/pull/301915)

amadaluzia reviewed (approved):
LGTM

niklaskorz reviewed (commented):
Could you add an entry to the `#other-notable-changes-sec-release-2505-notable-changes` section of `nixos/doc/manual/release-notes/rl-2505.section.md`? I think this would be nice to include in the release notes.

niklaskorz reviewed (approved):
Tests passing on x86_64-linux, also tried the module on my local system and it succesfully fulfills its purpose

Scrumplex commented on nixos/doc/manual/release-notes/rl-2505.section.md:
```diff
@@ -403,6 +403,8 @@
 
 - `gerbera` now has wavpack support.
 
+- A toggle has been added under the `users.users.<name>.enable` to allow toggling individual users conditionally. If set to false, the user account will not be created.
```
```suggestion
- A toggle has been added under `users.users.<name>.enable` to allow toggling individual users conditionally. If set to false, the user account will not be created.
```

NotAShelf replied:
This was meant to be `the ... option`. Whoops.

Fixing now.

Gerg-L reviewed (approved):
LGTM

NotAShelf commented:
Looks gerg to you I see...

nikstur commented:
I think this is good to go, you just need to resolve the conflicts for the release notes.

NotAShelf commented on nixos/tests/user-enable-option.nix:
```diff
@@ -0,0 +1,85 @@
+  };
+
+  testScript = ''
+    def check_user(username, status):
+      check_fn = f"id {username}"
+      if status == "success":
+        machine.succeed(check_fn)
+      elif status == "fail":
+        machine.fail(check_fn)
```
I don't think this is a difficult abstraction, but I'm open to changing it. I'll implement the change shortly.

NotAShelf commented:
Should be good now, I've fixed the merge conflict and rebased.

aanderse commented:
thanks so much for taking a look @nikstur, and thanks for your efforts pushing this through @NotAShelf!

nikstur commented:
@ofborg test userborn
@ofborg test userborn-mutable-users userborn-immutable-users
@ofborg test userborn-mutable-etc userborn-immutable-etc

@ofborg test user-enable-option mutableUsers

SuperSandro2000 commented on nixos/doc/manual/release-notes/rl-2505.section.md:
```diff
@@ -462,8 +462,11 @@
+
 - `ddclient` was updated from 3.11.2 to 4.0.0 [Release notes](https://github.com/ddclient/ddclient/releases/tag/v4.0.0)
 
+
```
```suggestion
```

SuperSandro2000 committed:
release-notes/25.05: drop extra added new line

SuperSandro2000 reviewed (approved):
lets give this a try :tada:

aanderse commented:
thanks!
